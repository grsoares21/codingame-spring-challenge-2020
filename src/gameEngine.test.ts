import { initDiffusionMatrices, getNewSignalMatrix } from "./gameEngine";

it(`should init highly attractive slightly attractive and repelling matrices`, () => {
  let map = getMap();
  let diffusionMatrices = initDiffusionMatrices(map);

  expect(
    prettyPrintNumberMatrix(
      diffusionMatrices.highlyAttractiveDiffusionMatrix[1][1]
    )
  ).toMatchSnapshot();
  expect(
    prettyPrintNumberMatrix(diffusionMatrices.repellingDiffusionMatrix[14][7])
  ).toMatchSnapshot();
  expect(
    prettyPrintNumberMatrix(
      diffusionMatrices.slightlyAttractiveDiffusionMatrix[16][7]
    )
  ).toMatchSnapshot();
});

it(`should get first combined signal matrix`, () => {
  let map = getMap();
  let diffusionMatrices = initDiffusionMatrices(map);
  let firstSignalMatrix = getNewSignalMatrix(
    map,
    null,
    diffusionMatrices,
    [
      { x: 14, y: 7 },
      { x: 16, y: 7 },
    ],
    [],
    []
  );
  expect(prettyPrintNumberMatrix(firstSignalMatrix)).toMatchSnapshot();
});

function prettyPrintNumberMatrix(matrix: number[][]): string {
  return matrix
    .map((line) =>
      line
        .map((num) => num.toFixed(1))
        .map((num) => (num === "-Infinity" ? "###" : num))
        .join(" ")
    )
    .join("\n");
}

function getMap(): string[][] {
  // prettier-ignore
  return [
    ['#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#'],
    ['#',' ','#',' ',' ',' ','#',' ',' ',' ','#',' ',' ',' ','#','#','#','#','#',' ',' ',' ','#',' ',' ',' ','#',' ',' ',' ','#',' ','#'],
    ['#',' ','#',' ','#','#','#','#','#',' ','#',' ','#','#','#','#','#','#','#','#','#',' ','#',' ','#','#','#','#','#',' ','#',' ','#'],
    ['#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#'],
    ['#','#','#',' ','#',' ','#',' ','#',' ','#',' ','#','#','#','#','#','#','#','#','#',' ','#',' ','#',' ','#',' ','#',' ','#','#','#'],
    [' ',' ',' ',' ','#',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ','#',' ',' ',' ',' '],
    ['#','#','#',' ','#',' ','#','#','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#','#','#',' ','#',' ','#','#','#'],
    ['#',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ','#'],
    ['#',' ','#','#','#',' ','#',' ','#',' ','#','#','#','#','#',' ','#',' ','#','#','#','#','#',' ','#',' ','#',' ','#','#','#',' ','#'],
    ['#',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ',' ',' ',' ',' ','#',' ',' ',' ',' ',' ','#'],
    ['#','#','#',' ','#',' ','#','#','#',' ','#',' ','#','#','#',' ','#',' ','#','#','#',' ','#',' ','#','#','#',' ','#',' ','#','#','#'],
    ['#',' ',' ',' ','#',' ',' ',' ',' ',' ','#',' ',' ',' ','#',' ','#',' ','#',' ',' ',' ','#',' ',' ',' ',' ',' ','#',' ',' ',' ','#'],
    ['#',' ','#',' ','#','#','#','#','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#',' ','#','#','#','#','#',' ','#',' ','#'],
    ['#',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ','#',' ',' ',' ',' ',' ',' ',' ','#',' ','#'],
    ['#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#','#']
  ];
}
