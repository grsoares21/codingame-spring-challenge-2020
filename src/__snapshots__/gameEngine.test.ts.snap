// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should get first combined signal matrix 1`] = `
"### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
### 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 3.9 3.5 0.0 ### ### ### ### ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 3.5 ### 7.8 ### ### ### ### ### ### ### ### ### 3.5 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 3.5 3.9 ### 8.7 7.8 3.9 3.5 0.0 0.0 0.0 0.0 0.0 3.5 3.9 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### 3.5 ### 7.8 ### 9.6 ### ### ### ### ### ### ### ### ### 7.8 ### 3.5 ### 0.0 ### 0.0 ### 0.0 ### ### ###
0.0 0.0 0.0 0.0 ### 0.0 ### 3.9 7.8 8.7 9.6 10.7 11.9 13.2 13.1 14.6 ### 13.2 11.9 10.7 9.6 8.7 7.8 3.9 3.5 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0
### ### ### 0.0 ### 0.0 ### ### ### 7.8 ### 9.6 ### 14.7 ### 16.2 ### 14.7 ### 11.9 ### 7.8 ### 3.5 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 3.5 3.9 7.8 8.7 ### 16.3 18.1 18.0 18.1 16.3 14.7 13.2 ### 3.9 3.5 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### ### ### 0.0 ### 0.0 ### 7.0 ### ### ### ### ### 16.2 ### 14.7 ### ### ### ### ### 3.5 ### 0.0 ### 0.0 ### ### ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 ### 0.0 7.0 7.7 8.6 9.6 10.6 11.8 13.1 14.6 ### 13.2 11.9 10.7 9.6 8.7 7.8 3.9 3.5 0.0 ### 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### ### ### 7.0 ### 8.6 ### ### ### 13.1 ### 11.9 ### ### ### 7.8 ### 3.5 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 7.7 7.0 0.0 ### 11.8 ### 10.7 ### 0.0 3.5 3.9 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 7.0 ### 7.0 ### 10.6 ### 9.6 ### 3.5 ### 3.5 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 7.7 8.6 9.6 8.6 8.7 7.8 3.9 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ###
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###"
`;

exports[`should init highly attractive slightly attractive and repelling matrices 1`] = `
"### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
### 10.0 ### 5.3 4.8 4.3 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### ### ### ### ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ###
### 9.0 ### 5.9 ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 8.1 7.3 6.6 5.9 5.3 4.8 4.3 3.9 3.5 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 5.9 ### 4.8 ### 3.9 ### 0.0 ### 0.0 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ###
3.9 4.3 4.8 5.3 ### 4.3 ### 3.5 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 3.5
### ### ### 4.8 ### 3.9 ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 3.5 3.9 4.3 ### 3.5 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ### 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ###
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###"
`;

exports[`should init highly attractive slightly attractive and repelling matrices 2`] = `
"### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
### 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### -1.9 -1.7 0.0 ### ### ### ### ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### -1.7 ### -2.2 ### ### ### ### ### ### ### ### ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 -1.7 -1.9 ### -2.4 -2.2 -1.9 -1.7 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### -1.7 ### -2.2 ### -2.7 ### ### ### ### ### ### ### ### ### -1.7 ### 0.0 ### 0.0 ### 0.0 ### 0.0 ### ### ###
0.0 0.0 0.0 0.0 ### 0.0 ### -1.9 -2.2 -2.4 -2.7 -3.0 -3.3 -3.6 -3.3 -3.6 ### -3.0 -2.7 -2.4 -2.2 -1.9 -1.7 0.0 0.0 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0
### ### ### 0.0 ### 0.0 ### ### ### -2.2 ### -2.7 ### -4.0 ### -4.0 ### -3.3 ### -2.7 ### -1.7 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 -1.7 -1.9 -2.2 -2.4 ### -4.5 -5.0 -4.5 -4.0 -3.6 -3.3 -3.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### ### ### 0.0 ### 0.0 ### -1.7 ### ### ### ### ### -4.0 ### -3.3 ### ### ### ### ### 0.0 ### 0.0 ### 0.0 ### ### ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 ### 0.0 -1.7 -1.9 -2.2 -2.4 -2.7 -3.0 -3.3 -3.6 ### -3.0 -2.7 -2.4 -2.2 -1.9 -1.7 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### ### ### -1.7 ### -2.2 ### ### ### -3.3 ### -2.7 ### ### ### -1.7 ### 0.0 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### -1.9 -1.7 0.0 ### -3.0 ### -2.4 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### -1.7 ### -1.7 ### -2.7 ### -2.2 ### 0.0 ### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### -1.9 -2.2 -2.4 -2.2 -1.9 -1.7 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ###
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###"
`;

exports[`should init highly attractive slightly attractive and repelling matrices 3`] = `
"### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###
### 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### ### ### ### ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 0.0 0.0 ### 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.3 ### ### ### ### ### ### ### ### ### 0.3 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.4 0.3 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.3 0.4 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### 0.0 ### 0.3 ### 0.4 ### ### ### ### ### ### ### ### ### 0.4 ### 0.3 ### 0.0 ### 0.0 ### 0.0 ### ### ###
0.0 0.0 0.0 0.0 ### 0.0 ### 0.0 0.3 0.4 0.4 0.5 0.5 0.6 0.7 0.7 ### 0.7 0.7 0.6 0.5 0.5 0.4 0.4 0.3 0.0 ### 0.0 ### 0.0 0.0 0.0 0.0
### ### ### 0.0 ### 0.0 ### ### ### 0.3 ### 0.4 ### 0.7 ### 0.8 ### 0.8 ### 0.7 ### 0.4 ### 0.3 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.3 0.4 ### 0.7 0.8 0.9 1.0 0.9 0.8 0.7 ### 0.4 0.3 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### ### ### 0.0 ### 0.0 ### 0.3 ### ### ### ### ### 0.8 ### 0.8 ### ### ### ### ### 0.3 ### 0.0 ### 0.0 ### ### ### 0.0 ###
### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.3 0.4 0.4 0.5 0.5 0.6 0.7 0.7 ### 0.7 0.7 0.6 0.5 0.5 0.4 0.4 0.3 0.0 ### 0.0 0.0 0.0 0.0 0.0 ###
### ### ### 0.0 ### 0.0 ### ### ### 0.3 ### 0.4 ### ### ### 0.7 ### 0.7 ### ### ### 0.4 ### 0.3 ### ### ### 0.0 ### 0.0 ### ### ###
### 0.0 0.0 0.0 ### 0.0 0.0 0.0 0.0 0.0 ### 0.4 0.3 0.0 ### 0.6 ### 0.6 ### 0.0 0.3 0.4 ### 0.0 0.0 0.0 0.0 0.0 ### 0.0 0.0 0.0 ###
### 0.0 ### 0.0 ### ### ### ### ### 0.0 ### 0.3 ### 0.3 ### 0.5 ### 0.5 ### 0.3 ### 0.3 ### 0.0 ### ### ### ### ### 0.0 ### 0.0 ###
### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ### 0.4 0.4 0.5 0.4 0.5 0.4 0.4 ### 0.0 ### 0.0 0.0 0.0 0.0 0.0 0.0 0.0 ### 0.0 ###
### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ###"
`;
